### DISCUSSION TITLE
Dynamically add / remove sources from fallback operator

### DISCUSSION THREAD

**Discussion by danbo** [savonet/liquidsoap#3394] [Q&A] [answered]
Hi,

When I look at dynamic sources, I see either request.dynamic or a way to dynamically create and destroy sources against different mount points ([here](https://github.com/savonet/liquidsoap/blob/2c51f1224eb596c2ce63ba6777838520ebf9c896/doc/content/dynamic_sources.md))

What I'm looking for is dynamically adding and removing sources from the fallback operator.

For example, right now I could have [requests, playqueue, jukebox] to support on the fly requests (requests), a place to push one or more playlists as a flat list (playqueue) and a preconfigured playlist (jukebox)

However, I was looking at not having playqueue flat but being able to dynamically add or remove playqueue1...n and have LS see those changes and respond accordingly.

ie, jukebox plays by default, a new source playqueue is added so LS either swaps to it immediately or after the current jukebox song finishes depending on track sensitive and while the playqueue is playing if a request is made, it jumps from playqueue to requests to play then back to playqueue unless it was removed at which point it falls back to jukebox.

This would allow easier management of the overall queue by simply adding and removing sources instead of x files/urls from one main queue. I know i could do this by having x statically defined queues but the idea is that I don't know how many I will need ahead of time.

I tried manipulating the fallback array while it was being used but i suspect LS takes a snapshot of it at the time of output creation.

.. unless there's a way to do this now that I'm missing.

**Comment by smimram** [comment, marked as answer]
I have just modified `native.fallback` (the alternative implementation of `fallback` using `source.dynamic` so that it accepts getters for sources: you can now pass a reference for the list of source. The PR is https://github.com/savonet/liquidsoap/pull/3398 and it shall be merged once the tests pass.

You can also have a look at the implementation of `native.fallback` which is basically what you want here. Suppose that you have `sources`, which is a reference to a list of sources (which you can thus modify dynamically). You can then create a source `s` which will play the first available one by
```
sources = ref([sine()]) # can be modified later on
def ready_sources() = list.find(default=fail, source.is_ready, !sources) end
s = source.dynamic(ready_source)
```

**Reply by Moonbase59** [reply]
@smimram With the new syntax, should it not be `sources()` instead of `!sources`?

**Reply by smimram** [reply]
> @smimram With the new syntax, should it not be `sources()` instead of `!sources`?

Yes! Still not used to the new syntax:)

**Comment by danbo** [comment]
@smimram. I can't seem to get this to work using a function (it does however work if I'm using the set function per the book's example)

Here's the sample:
```
dynamic_sources = ref([])
noSource = source.fail(id="nosource")
def ready_sources() = list.find(default=noSource, source.is_ready, dynamic_sources()) end

jukebox = source.dynamic(id="jukebox", track_sensitive=true, resurection_time=5., ready_sources)

def addSource(req, resp) =
  print("add")
  st = single(id="test", "./temp.mp3")
  dynamic_sources.set(list.add(st, dynamic_sources()))
  print(dynamic_sources())
  print("is source ready? #{st.is_ready()}")
  resp.data("The request was made!\n")
end

def delSource(req, resp) =
  print("del")
  st = list.hd(dynamic_sources())
  dynamic_sources.set(list.remove(st, dynamic_sources()))
  print(dynamic_sources())
  resp.data("The request was made!\n")
end

harbor.http.register(port=11111, method="GET", "/addSource", addSource)
harbor.http.register(port=11111, method="GET", "/delSource", delSource)

radio = mksafe(jukebox)
output.dummy(fallible=true, radio)
```

When calling ```curl http://localhost:11111/addSource``` I see that the single source is added to the list correctly but is not ready (not sure why) so it doesn't get returned by the find.

It doesn't matter if it's a file or a remote url. I don't see a resolve() function like on a request to force init/resolution.

```
...
2023/09/26 04:34:38 [mksafe:3] Switch to safe_blank.
add
[<source#test>.{time=<fun>, fallible=true, skip=<fun>, seek=<fun>, is_active=<fun>, is_up=<fun>, log={level=<fun>.{set=<fun>}}, self_sync=<fun>, duration=<fun>, elapsed=<fun>, remaining=<fun>, on_track=<fun>, on_shutdown=<fun>, on_wake_up=<fun>, on_metadata=<fun>, last_metadata=<fun>, buffered=<fun>, is_ready=<fun>, id=<fun>}]
is source ready? false
^C2023/09/26 04:35:11 [main:3] Shutdown signal received.
```

docker.io/savonet/liquidsoap:rolling-release-v2.2.x
Liquidsoap 2.2.2+git@80c0621ef

I also tried this with a recent version and same issue: Liquidsoap 2.2.2+git@ff1a04c38

**Reply by danbo** [reply]
So it looks like doing a ```list.find``` on ```source.is_ready``` isn't working because the source appears to be lazy loaded and that doesn't happen until it is referenced but of course I am trying to determine whether or not to reference it.

I was able to confirm this by simply by a matching source name instead, ie ```fun (s) -> s.id() == "test"```. At this point the source found and loaded as expected and was playing out of  ```dynamic_sources```

So then I either need a different condition for knowing whether or not to select the source or force load the source.

- I wasn't able to find a property on a source to determine if it has been played (atleast once)
- I wasn't able to find a method on a source to allow me to force load it (aside from ```request.resolve``` which is for a queue, ie on a playlist source but not on a single source.

Please let me know if I'm missing something.

Is there a plan to be able to dynamically add and remove sources from a fallback operator vs swapping the contents of a dynamic source / placeholder? This would allow LS to track play state across the sources like it does for existing operators and not require a find function / separate tracking.
